### PR TITLE
I894 detect bad accounts

### PR DESCRIPTION
**Pull request by danfinlay**
Detects bad accounts when migrating an old vault, adds them to `state.metamask.lostAccounts`.

This can then be added by the UI to show a warning, and we should add an action to clear those lostAccounts, too.

Still requires the UI & actions, but since those branches need to live together, one needs to be merged to `dev` before the other is done, hence this PR.

### CODE REVIEW

File: app/scripts/lib/config-manager.js [hunk 1]
```diff
@@ -432,3 +432,15 @@ ConfigManager.prototype.setGasMultiplier = function (gasMultiplier) {
...
+
+ConfigManager.prototype.setLostAccounts = function (lostAccounts) {
+  var data = this.getData()
+  data.lostAccounts = lostAccounts
+  this.setData(data)
+}
+
+ConfigManager.prototype.getLostAccounts = function () {
+  var data = this.getData()
+  return ('lostAccounts' in data) && data.lostAccounts || []
```
Comment (kumavis) [inline app/scripts/lib/config-manager.js]: i think you can drop `('lostAccounts' in data)`

File: app/scripts/lib/idStore-migrator.js [hunk 2]
```diff
@@ -26,11 +28,10 @@ module.exports = class IdentityStoreMigrator {
     return new Promise((resolve, reject) => {
       this.idStore.submitPassword(password, (err) => {
         if (err) return reject(err)
-        try {
-          resolve(this.serializeVault())
-        } catch (e) {
-          reject(e)
-        }
+        const serialized = this.serializeVault()
+        this.checkForErrors(serialized)
```
Comment (kumavis) [inline app/scripts/lib/idStore-migrator.js]: consider using [`denodeify`](https://github.com/matthew-andrews/denodeify) for clean callback-promise interface

File: app/scripts/lib/idStore-migrator.js [hunk 3]
```diff
@@ -45,6 +46,28 @@ module.exports = class IdentityStoreMigrator {
     }
   }
 
+  checkForErrors (serialized) {
```
Comment (kumavis) [inline app/scripts/lib/idStore-migrator.js]: consider renaming to `checkForLostAccounts` or something

File: test/unit/idStore-migration-test.js [hunk 4]
```diff
@@ -79,33 +83,9 @@ describe('IdentityStore to KeyringController migration', function() {
       keyringController.configManager.setWallet('something')
       const state = keyringController.getState()
       assert(state.isInitialized, 'old vault counted as initialized.')
+      console.dir(state)
```
Comment (kumavis) [inline test/unit/idStore-migration-test.js]: remove

File: app/scripts/keyrings/hd.js [hunk 5]
```diff
@@ -38,10 +38,10 @@ class HdKeyring extends EventEmitter {
     }
 
     if ('numberOfAccounts' in opts) {
```
Comment (kumavis) [inline app/scripts/keyrings/hd.js]: I reccomend using `if (opts.numberOfAccounts) {...}` bc it tells us if that value is truthy, as opposed to just present ( can be undefined and still present )
```js
z={a:undefined}; 'a' in z //=> true
```

**Review by kumavis — CHANGES_REQUESTED** [review]

**Review by 2-am-zzz — APPROVED** [review]
:+1:

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
L 👀 KS G👀 D

**Comment by kumavis** [comment]
( but notice the test is timing out -- maybe a promisify issue )

**Pull request merged**